pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,233 @@
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// reconcileSingletonByBdg goes over the binding-covered workload objects, decides sync or not for each of the workload objects,
+// then maintains their statuses based on the decisions.
+func (c *Controller) reconcileSingletonByBdg(ctx context.Context, bdgName string) error {
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Reconciling singleton status due to binding change", "binding", bdgName)
+
+	binding, err := c.wdsKsClient.ControlV1alpha1().Bindings().Get(ctx, bdgName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("could not get binding %s from cache", bdgName)
+	}
+
+	sync := make(map[util.ObjectIdentifier]bool)
+	if err = c.checkSingletonWorkloadInBinding(ctx, *binding, sync, true); err != nil {
+		return err
+	}
+
+	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, bdg := range allBdgs.Items {
+		if bdg.Name == binding.Name {
+			continue
+		}
+		if err = c.checkSingletonWorkloadInBinding(ctx, bdg, sync, false); err != nil {
+			return err
+		}
+	}
+
+	for wObjIdentifier, v := range sync {
+		if err := c.reconcileSingletonWObj(ctx, wObjIdentifier, v); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Controller) reconcileSingletonWObj(ctx context.Context, wObjID util.ObjectIdentifier, sync bool) error {
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Reconciling singleton workload object", "resource", wObjID.Resource, "objectName", wObjID.ObjectName)
+
+	if !sync {
+		emptyStatus := make(map[string]interface{})
+		logger.V(4).Info("Cleaning up singleton status", "resource", wObjID.Resource, "objectName", wObjID.ObjectName)
+		return updateObjectStatus(ctx, &wObjID, emptyStatus, c.listers, c.wdsDynClient)
+	}
+
+	list, _ := c.workStatusLister.ByNamespace("").List(labels.Everything())
+	for _, obj := range list {
+		wsRef, err := runtimeObjectToWorkStatusRef(obj)
+		if err != nil {
+			return err
+		}
+		sourceObjID := wsRef.SourceObjectIdentifier
+		if sourceObjID != wObjID {
+			continue
+		}
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return err
+		}
+		if status == nil {
+			return nil
+		}
+		logger.V(4).Info("Updating singleton status", "workstatus", wsRef.Name, "wecName", wsRef.WECName)
+		return updateObjectStatus(ctx, &wsRef.SourceObjectIdentifier, status, c.listers, c.wdsDynClient)
+	}
+	return nil
+}
+
+// reconcileSingletonByWs decides sync or not for the corresponding workload object,
+// then maintains its status based on the decision.
+func (c *Controller) reconcileSingletonByWs(ctx context.Context, ref singletonWorkStatusRef) error {
```

Comment:
I suppose that the "Ws" at the end of this method name is supposed to stand for "WorkStatus". If I have that right then I think that "WS" would do the job better.

Reply:
Addressed in #2409.